Docs: update environment design (#2199)
oam-dev/kubevela: design/vela-core/environment.md

# Making Environments For Application Deployment

## Background

> A application development team usually needs to initialize some shared environments for developers to deploy applications to.

Reviewer: ```suggestion
An application development team usually needs to initialize some shared environments for developers to deploy applications to.
```

For example, a team would initialize two environments, a *dev* environment for testing applications,
and a *prod* environment for running applications to serve live traffic.

The environment is a logical concept grouping common resources that many applications would depend on. Below are a list of cases:

> - Kubernetes clusters. These include not only existing clusters but also new clusters to create during environment setup.

Reviewer: ```suggestion
- Kubernetes clusters. These include not only existing clusters but also new clusters to be created during environment setup.
```

Author: 这个原先是正确的语法。。。

- Admin policies. Production environment would usually set global policies for all application deployments,
  like chaos testing, SLO requirements, security scanning, misconfiguration detection.

> - System components. That include installing system level Operators & CRDS like FluxCD, KEDA, OpenKruise,

Reviewer: ```suggestion
- System components. That includes installing system level Operators & CRDS like FluxCD, KEDA, OpenKruise,
```

>   shared Namespaces, Observability (Prometheus, Grafana, Loki), Terraform Controller, KubeFlow Controller, etc.

Reviewer: ```suggestion
  Shared Namespaces, Observability (Prometheus, Grafana, Loki), Terraform Controller, KubeFlow Controller, etc.
```

Author: 这个在 markdown 里没换行呢。。。

Reviewer: Sorry.

- Shared services. Environments contains a variety of system services that are be shared by applications.
  These resources can include DB, cache, load balancers, API Gateways, etc.

In this doc, we will provide solutions and case studies on how to use KubeVela to make environments, initialize shared resources,
and compose application rollout across multiple environments.

## Using KubeVela to initialize Environment

> An environment is just a logical concept that groups a bundle of resources for multiple applications to share.

Reviewer: This meaning has been expressed in a line above https://github.com/oam-dev/kubevela/pull/2199/files#diff-b42a63b4159dee9cf6945f592072aa04d2fb83b18a52c85ef3339f9df1f1bf24R9.

Author: Well. This sentence is good enough as the beginning of the section.

To make an environment, we first need to initialize the shared resources.

When initializing an environment, there is one additional requirement:

- The ability to describe dependencies between environment modules. For example, both *dev* and *prod* environments
  would depend on a common module that installs operators both *dev* and *prod* would use.

> To setup shared resources and define dependencies, we propose to use Application with two Workflow Steps

Reviewer: In the following demo, there is no workflow.

Author: What? It has two steps...

```
  workflow:
  - name: wait-dependencies
    # depends-on-app step will wait until the dependent app's status turns to Running phase
    type: depends-on-app
    properties:
      name: <application name>
      namespace: <application namespace>
      
  - name: apply-self
    # apply-application will apply all components of this Application per se
    type: apply-application
```

to achieve the requirements. An example looks like below:

```yaml
apiVersion: core.oam.dev/v1beta1

> kind: Application

Reviewer: I am still curious when an application can be called an environment? I should have missed something important.

Author: We will focus on how to use KubeVela to support environment-based application rollout.
We don't call an Application an environment, but focus on how we use KubeVela to support various scenarios.

metadata:
  name: setup-dev-env
spec:
  # Using Application to deploy resources that makes up an environment

  components:
  - name: <env-component-name>
    type: <env-component-type>
    properties:
      <env-component-props>

  policies:
  - name: <env-policy-name>
    type: <env-policy-type>
    properties:
      <env-policy-props>

  workflow:
  - name: wait-dependencies
    # depends-on-app step will wait until the dependent app's status turns to Running phase

>     type: depends-on-app

Reviewer: Is `workflow` in some application?

For this type, do all components of the application have to be ready first? Why not waiting for a component?

Author: Waiting for an app is good enough for initial use cases.
But definitely we can add more complex operations like waiting for a component.
We can wait until more use cases pops up and implement them later.

    and finally deploys the Application to the cluster.

## Implementation Plan

We will discuss the details of implementation plan in this section.

## Considerations

## Appendix

### Creating Kubernetes cluster

In this section, we will show how to create a Kubernetes cluster on Alibaba Cloud using Terraform.

We will first setup Terraform Alibab provider:

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: terraform-alibaba
  namespace: vela-system
spec:
  components:
    - name: default
      type: raw
      properties:
        apiVersion: terraform.core.oam.dev/v1beta1
        kind: Provider
        metadata:
          namespace: default
        spec:
          provider: alibaba
          region: cn-hongkong
          credentials:
            source: Secret
            secretRef:
              namespace: vela-system
              name: alibaba-account-creds
              key: credentials
  workflow:
  - name: wait-dependencies
    type: depends-on-app
    properties:
      name: terraform
      namespace: vela-system
      
  - name: apply-self
    type: apply-application
```

Then we will create a Kubernetes cluster:

```yaml

> apiVersion: core.oam.dev/v1beta1

Reviewer 2: this demo should be fixed

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: managed-cluster
  namespace: vela-system
spec:
  components:
    - name: ack-worker
      type: alibaba-ack
      properties:
        writeConnectionSecretToRef:
          name: ack-conn
          namespace: vela-system
  workflow:
    steps:
      - name: wait-dependencies
        type: depends-on-app
        properties:
          name: terraform-alibaba
          namespace: vela-system

      - name: wait-dependencies
        type: depends-on-app
        properties:
          name: ocm-cluster-manager
          namespace: vela-system
          
      - name: terraform-ack
        type: create-ack
        properties:
          component: ack-worker
        outputs:
          - name: connInfo
            exportKey: connInfo

      - name: register-ack
        type: register-cluster
        inputs:
          - from: connInfo
            parameterKey: connInfo
        properties:
          # user should set public network address of APIServer
          hubAPIServer: {{ public network address of APIServer }}
          env: prod
          initNameSpace: default
          patchLabels:
            purpose: test
```

Reviewer: Should the second step name `wait-dependencies` be renamed?

  namespace: vela-system
spec:
    components:
      - name: ack-worker
        type: alibaba-ack
        properties:
          writeConnectionSecretToRef:
            name: ack-conn
            namespace: vela-system
    workflow:
      steps:
        - name: terraform-ack
          type: create-ack
          properties:
            component: ack-worker
          outputs:
            - name: connInfo
              exportKey: connInfo

        - name: register-ack
          type: register-cluster
          inputs:
            - from: connInfo
              parameterKey: connInfo
          properties:
            # user should set public network address of APIServer
            hubAPIServer: {{ public network address of APIServer }}
            env: prod
            initNameSpace: default
            patchLabels:
              purpose: test

  workflow:
  - name: wait-dependencies
    type: depends-on-app
    properties:
      name: terraform-alibaba
      namespace: vela-system

  - name: wait-dependencies
    type: depends-on-app
    properties:
      name: ocm-cluster-manager
      namespace: vela-system
      
  - name: apply-self
    type: apply-application

> ```

Reviewer 2: This demo should be modified
```
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: managed-cluster
  namespace: vela-system
spec:
  components:
    - name: ack-worker
      type: alibaba-ack
      properties:
        writeConnectionSecretToRef:
          name: ack-conn
          namespace: vela-system
  workflow:
    steps:
      - name: wait-dependencies
        type: depends-on-app
        properties:
          name: terraform-alibaba
          namespace: vela-system

      - name: wait-dependencies
        type: depends-on-app
        properties:
          name: ocm-cluster-manager
          namespace: vela-system

      - name: terraform-ack
        type: create-ack
        properties:
          component: ack-worker
        outputs:
          - name: connInfo
            exportKey: connInfo

      - name: register-ack
        type: register-cluster
        inputs:
          - from: connInfo
            parameterKey: connInfo
        properties:
          # user should set public network address of APIServer
          hubAPIServer: {{ public network address of APIServer }}
          env: prod
          initNameSpace: default
          patchLabels:
            purpose: test
```
